protoc-gen-gogo/generator/generator.go

```diff
@@ -3209,6 +3217,106 @@ func (g *Generator) generateCommonMethods(mc *msgCtx) {
 	g.P("var xxx_messageInfo_", mc.goName, " ", g.Pkg["proto"], ".InternalMessageInfo")
 }
 
+func (g *Generator) cloneGenericHeader(typeName, returnName string) {
+	g.P("func (m *", typeName, ") Clone() ", returnName, "{")
+	g.In()
+	g.P("if m == nil {")
+	g.In()
+	g.P("return nil")
+	g.endBlock()
+	g.P("cloned := new(", typeName, ")")
+	g.P("*cloned = *m")
+	g.P()
+}
+
+func (g *Generator) cloneGenericFooter() {
+	g.P("return cloned")
+	g.endBlock()
+}
+
+func (g *Generator) ifNotNilHeader(name string) {
+	g.P("if m.", name, "!= nil {")
+	g.In()
+}
+
+func (g *Generator) endBlock() {
+	g.Out()
+	g.P("}")
+}
+
+func (g *Generator) generateClone(mc *msgCtx, topLevelFields []topLevelField, mapFields map[string]*GoMapDescriptor) {
+	g.cloneGenericHeader(mc.goName, fmt.Sprintf("*%s", mc.goName))
+
+	for _, field := range topLevelFields {
+		switch typedField := field.(type) {
+		case *simpleField:
+			switch *typedField.protoField.Type {
+			case descriptor.FieldDescriptorProto_TYPE_MESSAGE:
+				// If repeated do x, otherwise do y
+				if *typedField.protoField.Label == descriptor.FieldDescriptorProto_LABEL_REPEATED {
+					if mapDesc, ok := mapFields[typedField.goName]; ok {
+						// Handle Map
+						g.ifNotNilHeader(typedField.goName)
+						g.P("cloned.", typedField.goName, "= make(", mapDesc.GoType, ", ", "len(m.", typedField.goName, ")", ")")
+						g.P("for k, v := range m.", typedField.goName, " {")
+						g.In()
+						if *mapDesc.ValueField.Type == descriptor.FieldDescriptorProto_TYPE_MESSAGE {
+							g.P("cloned.", typedField.goName, "[k] = v.Clone()")
+						} else {
+							g.P("cloned.", typedField.goName, "[k] = v")
+						}
+						g.endBlock()
+						g.endBlock()
+					} else {
+						// Handle Slice
+						g.ifNotNilHeader(typedField.goName)
+						g.P("cloned.", typedField.goName, "= make(", typedField.goType, ", len(m.", typedField.goName, ")", ")")
+						g.P("for idx, v := range m.", typedField.goName, " {")
+						g.In()
+						// All slices here are messages, otherwise, they are not of type message
+						g.P("cloned.", typedField.goName, "[idx] = v.Clone()")
+						g.endBlock()
+						g.endBlock()
+					}
+				} else {
+					g.ifNotNilHeader(typedField.goName)
+					switch typedField.goType {
```

Comment:
Uhh, if you need to special-case here, the above logic can't work - what if you have a repeated `google.protobuf.Timestamp` field (or a map with timestamp value)? It would be better to add `Clone()` methods to the well-known types. I think there are `Makefile`s for generating the respective sources?